In [278]:
import pandas as pd

# Read the CSV and Perform Basic Data Cleaning
**Drop specified columns with .drop()**  
**Drop null value columns and rows with .dropna()**  
**Reset index and drop original index with .reset_index(drop = True**

In [279]:
# read data from file 
dataDf = pd.read_csv("archive/cumulative.csv")
dataDf.head()

rowid     kepid kepoi_name   kepler_name koi_disposition koi_pdisposition  \
0      1  10797460  K00752.01  Kepler-227 b       CONFIRMED        CANDIDATE   
1      2  10797460  K00752.02  Kepler-227 c       CONFIRMED        CANDIDATE   
2      3  10811496  K00753.01           NaN  FALSE POSITIVE   FALSE POSITIVE   
3      4  10848459  K00754.01           NaN  FALSE POSITIVE   FALSE POSITIVE   
4      5  10854555  K00755.01  Kepler-664 b       CONFIRMED        CANDIDATE   

   koi_score  koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  ...  \
0      1.000              0              0              0  ...   
1      0.969              0              0              0  ...   
2      0.000              0              1              0  ...   
3      0.000              0              1              0  ...   
4      1.000              0              0              0  ...   

   koi_steff_err2  koi_slogg  koi_slogg_err1  koi_slogg_err2  koi_srad  \
0           -81.0      4.467           0.064          -0.096     0.927   
1           -81.0      4.467           0.064          -0.096     0.927   
2          -176.0      4.544           0.044          -0.176     0.868   
3          -174.0      4.564           0.053          -0.168     0.791   
4          -211.0      4.438           0.070          -0.210     1.046   

   koi_srad_err1  koi_srad_err2         ra        dec  koi_kepmag  
0          0.105         -0.061  291.93423  48.141651      15.347  
1          0.105         -0.061  291.93423  48.141651      15.347  
2          0.233         -0.078  297.00482  48.134129      15.436  
3          0.201         -0.067  285.53461  48.285210      15.597  
4          0.334         -0.133  288.75488  48.226200      15.509  

[5 rows x 50 columns]

In [282]:
# get all columns string type
df.loc[:, dataDf.dtypes == object].columns

Index(['kepoi_name', 'kepler_name', 'koi_disposition', 'koi_pdisposition',
       'koi_tce_delivname'],
      dtype='object')

In [284]:
#remove the uncalulated columns
dataDfNew = dataDf.copy().drop(columns=["rowid", 
                                        "kepid", 
                                        "kepoi_name", 
                                        "kepler_name", 
                                        "koi_pdisposition", 
                                        "koi_score", 
                                        "koi_tce_delivname"])
# Drop the null columns
dataDfNew = dataDfNew.dropna(axis='columns', how='all').dropna().reset_index(drop=True).copy()
# Drop the null rows
dataDfNew.sample(10)

koi_disposition  koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  \
1877       CONFIRMED              0              0              0   
2617       CONFIRMED              0              0              0   
5926  FALSE POSITIVE              0              0              1   
1898  FALSE POSITIVE              0              0              1   
2113       CONFIRMED              0              0              0   
1897       CANDIDATE              0              0              0   
6072       CANDIDATE              0              0              0   
4656       CANDIDATE              0              0              0   
6524  FALSE POSITIVE              0              0              1   
3646  FALSE POSITIVE              0              1              1   

      koi_fpflag_ec  koi_period  koi_period_err1  koi_period_err2  \
1877              0   25.082340         0.000121        -0.000121   
2617              0    4.623345         0.000017        -0.000017   
5926              1    0.864651         0.000007        -0.000007   
1898              0    2.493238         0.000009        -0.000009   
2113              0    9.549277         0.000005        -0.000005   
1897              0   11.964074         0.000057        -0.000057   
6072              0   48.813012         0.000497        -0.000497   
4656              0   11.937835         0.000174        -0.000174   
6524              0    0.503200         0.000003        -0.000003   
3646              1    0.787120         0.000002        -0.000002   

      koi_time0bk  koi_time0bk_err1  ...  koi_steff_err2  koi_slogg  \
1877   134.135680          0.004210  ...          -227.0      4.241   
2617   132.370830          0.002930  ...          -124.0      4.484   
5926   131.692390          0.006920  ...          -182.0      4.556   
1898   132.636330          0.002780  ...          -136.0      3.344   
2113   175.757183          0.000433  ...           -80.0      4.899   
1897   133.158040          0.005120  ...          -175.0      4.437   
6072   157.606380          0.007480  ...          -209.0      4.395   
4656   133.261200          0.012900  ...          -162.0      4.611   
6524   131.906270          0.004570  ...          -154.0      4.639   
3646   131.739390          0.001900  ...          -201.0      4.481   

      koi_slogg_err1  koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2  \
1877           0.180          -0.180     1.262          0.335         -0.274   
2617           0.030          -0.120     0.970          0.150         -0.053   
5926           0.031          -0.178     0.851          0.226         -0.075   
1898           0.336          -0.275     4.131          1.547         -1.702   
2113           0.060          -0.060     0.369          0.050         -0.056   
1897           0.101          -0.188     0.961          0.278         -0.139   
6072           0.066          -0.306     1.110          0.581         -0.120   
4656           0.037          -0.112     0.738          0.131         -0.056   
6524           0.053          -0.021     0.630          0.029         -0.064   
3646           0.150          -0.100     0.787          0.093         -0.103   

             ra        dec  koi_kepmag  
1877  284.97800  41.420158      13.838  
2617  291.84805  44.968220      12.520  
5926  289.97064  43.662239      14.901  
1898  297.11157  47.534691      13.962  
2113  284.92184  45.972401      14.486  
1897  280.31488  43.985710      15.361  
6072  287.91861  44.121300      15.861  
4656  296.65256  46.618809      15.954  
6524  286.17670  42.196529      14.155  
3646  295.28110  40.132858      14.964  

[10 rows x 41 columns]

In [285]:
# Count number of times each value in koi_disposition occurs
dataDfNew["koi_disposition"].value_counts()

FALSE POSITIVE    4358
CONFIRMED         2272
CANDIDATE         2114
Name: koi_disposition, dtype: int64

In [286]:
X = dataDfNew.rename(columns={"koi_disposition": "label"})

In [287]:
y = X["label"]
y_names = ['FALSE POSITIVE', 'CONFIRMED', 'CANDIDATE']
X.drop(columns=["label"], inplace=True)
print(X.shape, y.shape)

(8744, 40) (8744,)


# Pre-processing

* Preprocess the dataset prior to fitting the model.
* Perform feature selection and remove unnecessary features.
* Use `MinMaxScaler` to scale the numerical data.
* Separate the data into training and testing data.

In [288]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, MinMaxScaler
from tensorflow.keras.utils import to_categorical

In [290]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)
X_scaler = MinMaxScaler().fit(X_train)
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [291]:
# Step 1: Label-encode data set
label_encoder = LabelEncoder()
label_encoder.fit(y_train)
encoded_y_train = label_encoder.transform(y_train)
encoded_y_test = label_encoder.transform(y_test)

# Step 2: Convert encoded labels to one-hot-encoding
y_train_categorical = to_categorical(encoded_y_train)
y_test_categorical = to_categorical(encoded_y_test)

Train the Support Vector Machine

In [299]:
y_train_categorical
encoded_y_train
# y_train

array([2, 2, 0, ..., 2, 1, 2])

# Train the Support Vector Machine

In [323]:
# Support vector machine linear classifier
from sklearn.svm import SVC 
from sklearn.metrics import classification_report
model = SVC(kernel='linear')
model.fit(X_train_scaled, y_train)
prediction = model.predict(X_test_scaled)
print(classification_report(y_test, prediction,
                            target_names=['CANDIDATE', 'CONFIRMED', 'FALSE POSITIVE']))

                precision    recall  f1-score   support

     CANDIDATE       0.73      0.56      0.63       523
     CONFIRMED       0.68      0.80      0.73       594
FALSE POSITIVE       0.98      1.00      0.99      1069

      accuracy                           0.84      2186
     macro avg       0.80      0.78      0.79      2186
  weighted avg       0.84      0.84      0.83      2186



In [324]:
print(f"Training Data Score: {model.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {model.score(X_test_scaled, y_test)}")

Training Data Score: 0.8502592253735896
Testing Data Score: 0.838975297346752


In [325]:
# Calculate classification report using encoded values to determine if numerical values change SVC outcome.
model = SVC(kernel='linear')
model.fit(X_train_scaled, encoded_y_train)
prediction = model.predict(X_test_scaled)

print(classification_report(encoded_y_test, prediction,
                            target_names=['CANDIDATE', 'CONFIRMED', 'FALSE POSITIVE']))

                precision    recall  f1-score   support

     CANDIDATE       0.73      0.56      0.63       523
     CONFIRMED       0.68      0.80      0.73       594
FALSE POSITIVE       0.98      1.00      0.99      1069

      accuracy                           0.84      2186
     macro avg       0.80      0.78      0.79      2186
  weighted avg       0.84      0.84      0.83      2186



In [326]:
print(f"Training encoded Data Score: {model.score(X_train_scaled, encoded_y_train)}")
print(f"Testing encoded Data Score: {model.score(X_test_scaled, encoded_y_test)}")

Training encoded Data Score: 0.8502592253735896
Testing encoded Data Score: 0.838975297346752


# Hyperparameter Tuning
* Tune Model Parameters
* Use `GridSearchCV` to tune model parameters.
* Tune and compare at least two different classifiers.

In [327]:
# Create the GridSearch estimator along with a parameter object containing the values to adjust
from sklearn.model_selection import GridSearchCV
param_grid = {'C': [1, 5, 10, 50],
              'gamma': [0.0001, 0.0005, 0.001, 0.005]}
grid = GridSearchCV(model, param_grid, verbose=3)

In [328]:
grid.fit(X_train_scaled, y_train)

Fitting 5 folds for each of 16 candidates, totalling 80 fits
[CV] C=1, gamma=0.0001 ...............................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ................... C=1, gamma=0.0001, score=0.840, total=   0.5s
[CV] C=1, gamma=0.0001 ...............................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.4s remaining:    0.0s


[CV] ................... C=1, gamma=0.0001, score=0.857, total=   0.5s
[CV] C=1, gamma=0.0001 ...............................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.9s remaining:    0.0s


[CV] ................... C=1, gamma=0.0001, score=0.825, total=   0.5s
[CV] C=1, gamma=0.0001 ...............................................
[CV] ................... C=1, gamma=0.0001, score=0.856, total=   0.5s
[CV] C=1, gamma=0.0001 ...............................................
[CV] ................... C=1, gamma=0.0001, score=0.850, total=   0.5s
[CV] C=1, gamma=0.0005 ...............................................
[CV] ................... C=1, gamma=0.0005, score=0.840, total=   0.5s
[CV] C=1, gamma=0.0005 ...............................................
[CV] ................... C=1, gamma=0.0005, score=0.857, total=   0.5s
[CV] C=1, gamma=0.0005 ...............................................
[CV] ................... C=1, gamma=0.0005, score=0.825, total=   0.5s
[CV] C=1, gamma=0.0005 ...............................................
[CV] ................... C=1, gamma=0.0005, score=0.856, total=   0.5s
[CV] C=1, gamma=0.0005 ...............................................
[CV] .

[CV] .................. C=50, gamma=0.0001, score=0.880, total=   0.7s
[CV] C=50, gamma=0.0001 ..............................................
[CV] .................. C=50, gamma=0.0001, score=0.886, total=   0.7s
[CV] C=50, gamma=0.0001 ..............................................
[CV] .................. C=50, gamma=0.0001, score=0.865, total=   0.6s
[CV] C=50, gamma=0.0001 ..............................................
[CV] .................. C=50, gamma=0.0001, score=0.890, total=   0.8s
[CV] C=50, gamma=0.0001 ..............................................
[CV] .................. C=50, gamma=0.0001, score=0.902, total=   0.7s
[CV] C=50, gamma=0.0005 ..............................................
[CV] .................. C=50, gamma=0.0005, score=0.880, total=   0.7s
[CV] C=50, gamma=0.0005 ..............................................
[CV] .................. C=50, gamma=0.0005, score=0.886, total=   0.7s
[CV] C=50, gamma=0.0005 ..............................................
[CV] .

[Parallel(n_jobs=1)]: Done  80 out of  80 | elapsed:   45.9s finished


GridSearchCV(cv=None, error_score=nan,
             estimator=SVC(C=1.0, break_ties=False, cache_size=200,
                           class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='scale', kernel='linear', max_iter=-1,
                           probability=False, random_state=None, shrinking=True,
                           tol=0.001, verbose=False),
             iid='deprecated', n_jobs=None,
             param_grid={'C': [1, 5, 10, 50],
                         'gamma': [0.0001, 0.0005, 0.001, 0.005]},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=3)

In [329]:
# List the best parameters for this dataset
print(grid.best_params_)

{'C': 50, 'gamma': 0.0001}


In [330]:
print(grid.best_score_)

0.8845719149411174


In [331]:
# Make predictions with the hypertuned model
predictions = grid.predict(X_test_scaled)

In [332]:
# Calculate classification report
print(classification_report(y_test, predictions,
                            target_names=['CANDIDATE', 'CONFIRMED', 'FALSE POSITIVE']))

                precision    recall  f1-score   support

     CANDIDATE       0.85      0.65      0.73       523
     CONFIRMED       0.75      0.87      0.81       594
FALSE POSITIVE       0.98      1.00      0.99      1069

      accuracy                           0.88      2186
     macro avg       0.86      0.84      0.84      2186
  weighted avg       0.88      0.88      0.88      2186



In [333]:
print(f"Testing grid Data Score: {grid.score(X_test_scaled, y_test)}")

Testing grid Data Score: 0.8810612991765783


## Neural Network Learning

In [361]:
from keras.models import Sequential
from keras.layers import Dense
model = Sequential()
model.add(Dense(units=120, activation='relu', input_dim=40))
# model.add(Dense(units=120, activation='relu'))
model.add(Dense(units=3, activation='softmax'))

In [362]:
model.summary()

Model: "sequential_10"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_22 (Dense)             (None, 120)               4920      
_________________________________________________________________
dense_23 (Dense)             (None, 3)                 363       
Total params: 5,283
Trainable params: 5,283
Non-trainable params: 0
_________________________________________________________________


In [363]:
# Compile the model
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

In [364]:
# Fit the model to the training data
model.fit(
    X_train_scaled,
    y_train_categorical,
    epochs=100,
    shuffle=True,
    verbose=2
)

Epoch 1/100
 - 0s - loss: 0.6751 - accuracy: 0.7030
Epoch 2/100
 - 0s - loss: 0.4017 - accuracy: 0.8073
Epoch 3/100
 - 0s - loss: 0.3649 - accuracy: 0.8205
Epoch 4/100
 - 0s - loss: 0.3519 - accuracy: 0.8251
Epoch 5/100
 - 0s - loss: 0.3406 - accuracy: 0.8356
Epoch 6/100
 - 0s - loss: 0.3352 - accuracy: 0.8414
Epoch 7/100
 - 0s - loss: 0.3300 - accuracy: 0.8426
Epoch 8/100
 - 0s - loss: 0.3231 - accuracy: 0.8509
Epoch 9/100
 - 0s - loss: 0.3195 - accuracy: 0.8533
Epoch 10/100
 - 0s - loss: 0.3142 - accuracy: 0.8559
Epoch 11/100
 - 0s - loss: 0.3140 - accuracy: 0.8544
Epoch 12/100
 - 0s - loss: 0.3061 - accuracy: 0.8588
Epoch 13/100
 - 0s - loss: 0.3073 - accuracy: 0.8556
Epoch 14/100
 - 0s - loss: 0.3039 - accuracy: 0.8637
Epoch 15/100
 - 0s - loss: 0.3014 - accuracy: 0.8623
Epoch 16/100
 - 0s - loss: 0.3010 - accuracy: 0.8643
Epoch 17/100
 - 0s - loss: 0.2950 - accuracy: 0.8658
Epoch 18/100
 - 0s - loss: 0.2941 - accuracy: 0.8666
Epoch 19/100
 - 0s - loss: 0.2922 - accuracy: 0.8683
Ep

In [365]:
model_loss, model_accuracy = model.evaluate(
    X_test_scaled, y_test_categorical, verbose=2)
print(
    f"Normal Neural Network - Loss: {model_loss}, Accuracy: {model_accuracy}")

Normal Neural Network - Loss: 0.3254155432643524, Accuracy: 0.8902104496955872


#### Conclusion
The neural achieves have the more accuracy of 89% than the the Hypertuning(88%) and SVM model(83%). However with the Hypertuning grid search model achieves higher accuracy of 88% than Support Vector Machine. It would be interesting to see if/how-many epochs of the neural network it would take to achieve the same accuracy. 